exasol_script_languages_developer_sandbox/lib/deployer.py

```diff
@@ -0,0 +1,212 @@
+import collections
+
+from datetime import datetime
+
+from botocore.exceptions import ValidationError
+
+LOG = logging.getLogger(__name__)
+
+ChangeSetResult = collections.namedtuple(
+    "ChangeSetResult", ["changeset_id", "changeset_type"])
+
+
+class Deployer(object):
+
+    def __init__(self, cloudformation_client,
+                 changeset_prefix="slc-ci-setup-deploy-"):
+        self._client = cloudformation_client
+        self.changeset_prefix = changeset_prefix
+
+    def has_stack(self, stack_name):
+        """
+        Checks if a CloudFormation stack with given name exists
+
+        :param stack_name: Name or ID of the stack
+        :return: True if stack exists. False otherwise
+        """
+        try:
+            resp = self._client.describe_stacks(StackName=stack_name)
+            if len(resp["Stacks"]) != 1:
+                return False
+
+            # When you run CreateChangeSet on a a stack that does not exist,
+            # CloudFormation will create a stack and set it's status
+            # REVIEW_IN_PROGRESS. However this stack is cannot be manipulated
+            # by "update" commands. Under this circumstances, we treat like
+            # this stack does not exist and call CreateChangeSet will
+            # ChangeSetType set to CREATE and not UPDATE.
+            stack = resp["Stacks"][0]
+            return stack["StackStatus"] != "REVIEW_IN_PROGRESS"
+
+        except botocore.exceptions.ClientError as e:
+            # If a stack does not exist, describe_stacks will throw an
+            # exception. Unfortunately we don't have a better way than parsing
+            # the exception msg to understand the nature of this exception.
+            msg = str(e)
+
+            if "Stack with id {0} does not exist".format(stack_name) in msg:
+                LOG.debug("Stack with id {0} does not exist".format(
+                    stack_name))
+                return False
+            else:
+                # We don't know anything about this exception. Don't handle
+                LOG.debug("Unable to get stack details.", exc_info=e)
+                raise e
+
+    def create_changeset(self, stack_name, cfn_template,
+                         parameter_values, capabilities, role_arn,
+                         notification_arns, tags):
+        """
+        Call Cloudformation to create a changeset and wait for it to complete
+
+        :param stack_name: Name or ID of stack
+        :param cfn_template: CloudFormation template string
+        :param parameter_values: Template parameters object
+        :param capabilities: Array of capabilities passed to CloudFormation
+        :param tags: Array of tags passed to CloudFormation
+        :return:
+        """
+
+        now = datetime.utcnow().isoformat()
+        description = "Created by AWS CLI at {0} UTC".format(now)
+
+        # Each changeset will get a unique name based on time
+        changeset_name = self.changeset_prefix + str(int(time.time()))
```

Comment:
I guess, we only need one changeset, such that time is unique enough

Reply:
This is the code I copied from AWS CLI. I would prefer to keep it as is.

Reply:
Ah missed that. Then it's ok.